doc/designs/quic-design/quic-api.md

```diff
@@ -0,0 +1,752 @@
+ * will then populate the event queue accordingly and fill ev with the first
+ * event. For DTLS this could fulfil the role of DTLSv1_listen() and create a
+ * "new connection" event only when the cookie has been verified.
+ * If after the above the event queue is still empty and SSL_EVENT_TYPE_TICK is
+ * not masked out then it will populate the event queue with any expired timer
+ * events for s if it is non NULL, or any SSL objects otherwise.
+ * If shouldblock is 1 and no events are available this function call will block
+ * until an event is available, or the timeout value has expired. If shouldblock
+ * is 1 and timeout is NULL then the function will block indefinitely until an
+ * event is available. If shouldblock is 0 then then function will always return
+ * immediately.
+ */
+int SSL_EVENT_CTX_get_next_event(SSL_EVENT_CTX *ectx, SSL *s,
+                                 int shouldblock, struct tm *timeout,
+                                 SSL_EVENT *ev);
+
+/*
+ * Add a pending new connection to the new connection queue in the SSL_EVENT_CTX.
+ * This will result in an SSL_EVENT_TYPE_NEW_CONNECTION event being issued. A
+ * subsequent call to SSL_accept will accept the connection and configure the
+ * SSL object for it. Applications should only need to do this for TLS because we
+ * don't otherwise know about the arrival of a new connection. Even for TLS this
+ * is optional (new connections can just be handled as they are now without the
+ * use of SSL_EVENT_CTX if desired). For QUIC and DTLS these new connections
+ * can be detected automatically when SSL_EVENT_CTX_get_next_event() processes a
+ * packet.
+ */
+int SSL_EVENT_CTX_add_new_connection(SSL_EVENT_CTX *ectx, BIO *rbio, BIO *wbio);
+
+/*
+ * Construct and free a new SSL_EVENT object
+ */
+SSL_EVENT *SSL_EVENT_new(void);
+void SSL_EVENT_free(SSL_EVENT *ev);
+
+
+/* Gives an integer value representing the type of event that occurred */
+uint64_t SSL_EVENT_get_event_type(SSL_EVENT *event);
+
+/*
+ * Gets the stream associated with a given event. In the case of "new connection"
+ * this will be NULL. In case the of "new stream" or "connection close" it will
+ * be the "default" stream (i.e. first stream) for the connection
+ */
+SSL *SSL_EVENT_get_stream(SSL_EVENT *event);
+
+/*
+ * Works as normal if the SSL object has already had its BIOs configured.
+ * Otherwise it will check the new connection queue in the SSL_EVENT_CTX. If it
+ * finds a pending new connection then it will "accept" it, adding the BIOs and
+ * running the handshake. In the case of QUIC this will set up the SSL object as
+ * the "default" stream. Will return SSL_ERROR_WANT_READ in the event that no
+ * new conneciton is pending.
+ */
+__owur int SSL_accept(SSL *ssl);
+
+/*
+ * Works as normal for (D)TLS. For QUIC it is essentially the same but (as for
+ * SSL_accpet()) it will typically use the BIOs as set on the SSL_EVENT_CTX
+ * rather than having its own BIOs.
+ */
+__owur int SSL_connect(SSL *ssl);
```

Comment:
@hlandau wrote:
> This needs elaboration. How is the destination address specified? The example code also never seems to specify a destination.

This is also an existing function. For (D)TLS the underlying socket is "connected" so it is never necessary to specify the peer address - its implicit in the socket that the application created. For QUIC, since this is client side only, is there any reason why we can't do the same? The servername (for SNI purposes) has to be set separately via a call to `SSL_set_tlsext_host_name`. That's a bit nasty but is the way things work now - so no difference for QUIC.

Reply:
Not sure I understand. Maybe you could update the example code to show how the destination is specified?

Reply:
I mean the underlying socket has had the "connect" system call called for it:

````
       int connect(int sockfd, const struct sockaddr *addr,
                   socklen_t addrlen);
````
So libssl itself never needs to know the actual IP address (the `peer` argument in the `BIO_sendmmsg` call that libssl makes can just be NULL in this case).

Separate to that TLS (optionally) sends the servername in the ClientHello. That has to be set via a call to `SSL_set_tlsext_host_name`.

I'll expand out the example code a bit.

Reply:
Ah, right. I was looking at this with the assumption that the same UDP socket is involved in multiple connections. This is the case for a server but need not be the case for a client, so not strictly an issue.

Reply:
Exactly. `SSL_connect` is only ever called client side.

Reply:
Might be worth noting, this might not be tenable forever (beyond the MVP) when adding support for connection migration. We might want to figure out what kind of API implications connection migration might eventually have.

Reply:
Why does connection migration make a difference regarding whether or not we use connected sockets? Presumably connection migration will mean we have multiple sockets associated with the connection - but they can still be "connected".

Reply:
I meant that if an application provides sockets to OpenSSL having already called connect on them, that exposes this detail to the application. So what if during migration OpenSSL needs a socket with a different destination?

I've just checked and it looks like UDP sockets can have `connect` called on them multiple times to change their destination. We might want to check how portable this is. Of course you can also avoid using `connect` and call `sendto` every time.

So... it's the case that we might only need one socket, and can re-`connect` it if necessary without re-involving the application. It's still a little weird me that we would have the application do the `connect` directly on the socket the first time and then have OpenSSL changing that behind the scenes. Applications might find that surprising. Presumably OpenSSL will find out the originally set destination via `getpeername` since it will probably need to know at some point.

Reply:
I've now added an `SSL_connect_ex()` function which additionally takes a hostname and a destination address. This would enable the use of a socket in an "unconnected" state and would support the scenario where the server wishes to migrate to a preferred address.